### DISCUSSION TITLE
ESP32 stops subscribing after a while

### DISCUSSION THREAD

**Discussion by Bernard314** [bertmelis/espMqttClient#101] [Q&A] [answered]
I am using an esp32 to track my machine usage. it publishes the machine data using mqtt publish qos 1. i have also made the esp32 subscribe to a topic where i send in some inputs.

My issue is my code works perfectly fine but after a few hours it stops receiving messages. it still publishes the data so i know it is connected to mqtt without any problem, but why does it stop receiving these messages? 
My code is really simple just subscribe (i unsubscribe and then resubscribe after every 30 minutes to try to see it that fixes it also. didint help) and publishes.

Any help?

**Comment by bertmelis** [comment]
That would be very odd behaviour. If it is connected, it means it sends **and** receives PING/PONG messages.

Are you able to get debug output via Serial? --> set core debug log level to "verbose".
Can you share your code?
Anything special in the log of the broker?

**Reply by Bernard314** [reply]
ok so I tried again and now it works. So basically, sometimes it works and sometimes it just doesn't. I'm not sure if it's my Wi-Fi as it was working while I checked. maybe i need a way to implement qos level 1 so messages reach my esp reliably. Seems like a network issue though since i have noticed some messages that i have arranged to be published at certain times of the day reach the database up to an hour later than they should. 

I will look up how can i send qos level 1 messages through AWS. Currently, my messages go through lambda and are set at qos level 1 but that doesn't work. Maybe i need to do something else.

**Reply by bertmelis** [reply]
I can't help you with AWS. I have zero knowledge.

But it sparked my interest... Something like this: https://stackoverflow.com/questions/37810289/how-can-i-publish-to-a-mqtt-topic-in-a-amazon-aws-lambda-function or https://stackoverflow.com/questions/67441126/send-mqtt-messages-from-aws-lambda-to-my-iot-device?noredirect=1&lq=1

**Reply by Bernard314** [reply]
yeah its so weird i cant figure it out. it kinda stopped receiving again and then after some time it started receiving again. even though i figured out how to publish qos 1 messages, the ones that weren't received when i sent them, were not received at all, but the ones i sent later were received properly. i have a line of code that tells the esp to unsubscribe and then subscribe again after every 30 minutes (i did this just as an attempt to keep the connection fresh). maybe that could be causing the issue? i'll try removing that line and see what happens.

**Reply by bertmelis** [reply]
Make sure you get an subscribe confirmation (a positive confirmation that is, of course).
Could also be that your esp has reconnected with with a clean session so the subscription has to be renewed.

**Reply by Bernard314** [reply]
how can i get a subscribe confirmation? what does that do?

yeah you're right my esp is set to reconnect with a clean session. is that a problem? i think the connection was interrupted due to wifi issues, so when it reconnected, it did so with a clean session. what problem does that cause?

**Reply by bertmelis** [reply]
There is an onSubscribe callback. It has the return code as argument.

> espMqttClient& onSubscribe(espMqttClientTypes::OnSubscribeCallback callback)
>
> Add a subscribe acknowledged event handler. Function signature: void(uint16_t packetId, const espMqttClientTypes::SubscribeReturncode* returncodes, size_t len)

The return code is one of these: https://github.com/bertmelis/espMqttClient/blob/main/src/TypeDefs.h#L33
The callback has as arguments a pointer to an array of return codes with size `len`. (You can subscribe to multiple topics on one subscribe call). It gives you the maximum QoS with which messages are sent to the client.

Wrt reconnecting: when clean session is set to true (or the onConnect handler shows no session is present), you will have to resubscribe.

**Reply by Bernard314** [reply]
> There is an onSubscribe callback. It has the return code as argument.

Oh that's what you meant. Yeah i'll try that.

> Wrt reconnecting: when clean session is set to true (or the onConnect handler shows no session is present), you will have to resubscribe.

Ok i tried setting the setCleanSession to false this time. Maybe that was causing the problem

Here's my code (the pub and sub part at least)

```
#include <WiFi.h>
#include <espMqttClient.h>
#include "ArduinoJson.h"
#include "time.h"
#include <Arduino.h>
#include <AsyncTCP.h>
#include <ESPAsyncWebServer.h>
#define WIFI_SSID "username"
#define WIFI_PASSWORD "password"
#define MQTT_PORT 8883
#define MQTT_USER "username"
#define MQTT_PASS "password"


const char* awsEndpoint = "";

// xxxxxxxxxx-certificate.pem.crt
const char certificate_pem_crt[] = \

"-----BEGIN CERTIFICATE-----\n"\
"-----END CERTIFICATE-----\n";

// xxxxxxxxxx-private.pem.key
const char private_pem_key[] = \
"-----BEGIN RSA PRIVATE KEY-----\n"\
"-----END RSA PRIVATE KEY-----\n";


const char rootCA[] = \

"-----BEGIN CERTIFICATE-----\n" \
"-----END CERTIFICATE-----\n";

espMqttClientSecure mqttClient;

const char* clientID = "client_id";  
const char* topicchar = "current";
const char* topicchar_rep = "current_rep";    
const char* topicchar_check = "current_check";
char* output1_check = "received";

char* output1;
char output[350];


/*(1)functions for mqtt start*/
void connectToWiFi() {
  // Serial.println("Connecting to Wi-Fi...");
  WiFi.begin(WIFI_SSID, WIFI_PASSWORD);
}
void connectToMqtt() {
  // Serial.println("Connecting to MQTT...");             
  mqttClient.connect();
}

void WiFiEvent(WiFiEvent_t event) {
  Serial.printf("[WiFi-event] event: %d\n", event);
  switch(event) {
    case SYSTEM_EVENT_STA_GOT_IP:
      // Serial.println("WiFi connected");
      // Serial.println("IP address: ");
      // Serial.println(WiFi.localIP());
      connectToMqtt();
      break;
    case SYSTEM_EVENT_STA_DISCONNECTED:
      // Serial.println("WiFi lost connection");
      connectToWiFi();
      break;
    default:
      break;
  }
}


/*(1)functions for mqtt end*/


void onMqttConnect(bool sessionPresent) {
  Serial.println("Connected to MQTT.");
  Serial.print("Session present: ");
  Serial.println(sessionPresent);
  mqttClient.subscribe("pubtoskoda", 1);
  Serial.print("Subscribing at QoS 0, packetId: ");

  
}


void onMqttMessage(const espMqttClientTypes::MessageProperties& properties, const char* topic, const uint8_t* payload, size_t len, size_t index, size_t total) {
  (void) payload;
  StaticJsonDocument<200> pubdoc;
  DeserializationError error = deserializeJson(pubdoc, payload, len);

  // Check if parsing was successful
  if (error) {
    Serial.print("Parsing failed: ");
    Serial.println(error.c_str());
    return;
  }

	//confirmation that message has been received
  mqttClient.publish(topicchar_check, 0, false, output1_check);
}

void setup() {
  Serial.begin(115200);

  WiFi.onEvent(WiFiEvent);
  WiFi.setAutoConnect(false);
  WiFi.setAutoReconnect(true);
  mqttClient.setCACert(rootCA);
  mqttClient.setCertificate(certificate_pem_crt);
  mqttClient.setPrivateKey(private_pem_key);
  mqttClient.setClientId(clientID);
  mqttClient.onConnect(onMqttConnect);  
  mqttClient.onMessage(onMqttMessage);
  mqttClient.setServer(awsEndpoint, MQTT_PORT);
  mqttClient.setCleanSession(false);   
  
  delay(2500);
  connectToWiFi();
  


  mqttClient.connect();
  // delay(10000);

}


void loop() {
    //disconnect and connect after ecery 30 minutes to make sure everything is ok
   //i removed this commented out part a few hours ago 
/*
  if (15 minutes have passed) {


   if(res == 0){
      res++;
    }else if(res == 1){
      mqttClient.unsubscribe("pubtoskoda");
      mqttClient.subscribe("pubtoskoda", 1);
      res = 0;
      }
    if(rec == 0){
      rec++;
    }else if(rec == 5){
      if(mqttClient.disconnected()){
        mqttClient.connect();
      }
    }
*/
    if (the current time is between 7 am and 7 30 am){
      mqttClient.publish(topicchar_rep, 1, false, output1_rep);
    }
    if (the current time is between 7 pm and 7 30 pm){
        mqttClient.publish(topicchar_rep, 1, false, output1_rep);
        }
        
        }
        
}
```

**Reply by Bernard314** [reply]
I think it is some issue with my wifi. After changing the qos to 1 i am able to receive the messages but sometimes they reach after 30 minutes, sometimes they reach instantly. nevertheless, i'll still research how can i reduce this time or if it is actually an issue with my wifi or not.
thank you!

**Reply by bertmelis** [reply]
Always happy to help.

(I will behaving limited connectivity during the coming weeks though.)

**Reply by Bernard314** [reply]
ok so apologies but i checked again. i set up a callback evrytime my esp disconnects and found out that it disconnects very frequently after the one hour mark. im now sure it's not my wifi thats causing the problem as it has been reliable for weeks now.

**Reply by Bernard314** [reply]
here is the debug output

```
[  8118][D][ssl_client.cpp:282] start_ssl_client(): Protocol is TLSv1.2 Ciphersuite is TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256
[  8118][D][ssl_client.cpp:284] start_ssl_client(): Record expansion is 29
Connected to MQTT.
Session present: 1
Subscribing at QoS 0, packetId: Received job: Ho from AW IoT console
Disconnected from MQTT: 7.
[2862113][D][ssl_client.cpp:282] start_ssl_client(): Protocol is TLSv1.2 Ciphersuite is TLS-ECDHE-RSA-WITH-AES-128-GCM-SHA256
[2862114][D][ssl_client.cpp:284] start_ssl_client(): Record expansion is 29
Connected to MQTT.
Session present: 1
Subscribing at QoS 0, packetId: [2890139][D][WiFiGeneric.cpp:1035] _eventCallback(): Arduino Event: 5 - STA_DISCONNECTED
[2890140][W][WiFiGeneric.cpp:1057] _eventCallback(): Reason: 200 - BEACON_TIMEOUT

```


and sometime later this 
```

disconnected from MQTT: 7.
[11486499][E][WiFiGeneric.cpp:1578] hostByName(): DNS Failed for m6qs-ats.iot.us-east-1.amazonaws.com
Disconnected from MQTT: 7.
[11611492][E][WiFiClientSecure.cpp:144] connect(): start_ssl_client: -1
Disconnected from MQTT: 7.
[11618503][E][WiFiGeneric.cpp:1578] hostByName(): DNS Failed for m6qs-ats.iot.us-east-1.amazonaws.com
Disconnected from MQTT: 7.
[11625509][E][WiFiGeneric.cpp:1578] hostByName(): DNS Failed for qs-ats.iot.us-east-1.amazonaws.com
Disconnected from MQTT: 7.
```

**Reply by bertmelis** [reply]
Would it be possible to get a log from the broker?

Somehow the broker disconnects. And because there is no disconnect package, the reason is unknown to the client.

**Reply by Bernard314** [reply]
these are the 2 main ones. the first one is most popular

```
{
    "timestamp": "2023-07-13 06:11:41.775",
    "logLevel": "INFO",
    "traceId": "d",
    "accountId": "71436",
    "status": "Success",
    "eventType": "Disconnect",
    "protocol": "MQTT",
    "clientId": "vtl_client_id",
    "principalId": "8",
    "sourceIp": "9",
    "sourcePort": 62,
    "disconnectReason": "MQTT_KEEP_ALIVE_TIMEOUT"
}


{
    "timestamp": "2023-07-13 04:42:40.522",
    "logLevel": "INFO",
    "traceId": "8",
    "accountId": "436",
    "status": "Success",
    "eventType": "Disconnect",
    "protocol": "MQTT",
    "clientId": "vtl_client_id",
    "principalId": "b8",
    "sourceIp": "9",
    "sourcePort": 664,
    "disconnectReason": "CONNECTION_LOST"
}

```

**Reply by bertmelis** [reply]
Thank you. Very useful information.
AWS seems to be very precise when it comes to keepalive. What value have you set?

**Reply by Bernard314** [reply]
just the default

**Reply by bertmelis** [reply]
You could try with changing line https://github.com/bertmelis/espMqttClient/blob/main/src/MqttClient.cpp#L451:

from
```cpp
  if (!_pingSent &&
      ((currentMillis - _lastClientActivity > _keepAlive) ||
       (currentMillis - _lastServerActivity > _keepAlive))) {
```

to (for example)
```cpp
  if (!_pingSent &&
      ((currentMillis - _lastClientActivity > _keepAlive / 2) ||
       (currentMillis - _lastServerActivity > _keepAlive))) {
```

Doing so will force the client to send a PING in half the interval instead of at the end of the keepalive window and may improve your situation.

**Reply by bertmelis** [reply]
another option is to set the keepalive higher then the expected frequency you expect to send packets from the client (or receive qos >0 packets).

So if you send a packet every minute, set keepalive to 75 seconds or so.

**Reply by Bernard314** [reply]
ok so I ran a few tests. some at my place of work (where I intend to use my project) and some at home and now i can say for sure it has something to do with the Wi-Fi at work. At home I ran the test for 3 hours (twice) and have faced absolutely no interruptions in both cases. while at work, issues start showing up after about an hour. 

this is the log of my aws mqtt broker, of when my client connected subscribed and disconnected. All the disconnected logs mention "MQTT_KEEP_ALIVE_TIMEOUT"

<img width="292" alt="image" src="https://github.com/bertmelis/espMqttClient/assets/92743799/a0aac5bf-867c-4b48-ba01-4595c6b626f8">

i increased the keepalive to 120 seconds. i send a pack every few hours. can i keep the timeout time that high? like i only send about 10 packets a day (i need a consistent connection because I am subscribing to receive some inputs randomly)

My home Wi-Fi is much better than my work Wi-Fi, but I don't think it's that bad. (It is pretty consistent, but it does have several cameras linked to its ethernet). But like mqtt is supposed to be lightweight. I am going to check some properties and investigate my work Wi-Fi tomorrow.

**Reply by bertmelis** [reply]
I can change the code but it's actually AWS that is messing up here. The spec says the broker **should** disconnect after 1.5x keepalive. I is allowed to disconnect whenever it wants to.
The client sends a ping on 1x keepalive.

Keepalive can be set to several hours but you lose "responsiveness". You don't know if the client is still online until keepalive expires.

I still wonder why you're the first one that suffers this issue. I also need a closer look into the log timestamps. Does it disconnect exactly at keepalive later?

**Reply by bertmelis** [reply]
Ok, I did some searching and reading. I don't know what is true as I don't use AWS myself. AWS is respecting the 1,5 keepalive according some Google sources but the minimum value is 30sec. So it might be the client itself that is disconnecting because it thinks keepalive is set at 15.

Can you confirm the 120s works on both locations?

**Reply by bertmelis** [reply]
Source of the keepalive limitation on AWS is here: https://docs.aws.amazon.com/iot/latest/developerguide/mqtt.html#mqtt-differences

**Comment by Bernard314** [comment]
> Ok, I did some searching and reading. I don't know what is true as I don't use AWS myself. AWS is respecting the 1,5 keepalive according some Google sources but the minimum value is 30sec. So it might be the client itself that is disconnecting because it thinks keepalive is set at 15.
> 
> Can you confirm the 120s works on both locations?
Yeah that's the thing. The code even with default value works perfectly at my home wifi. But not (even with 120s) at my work wifi. I'll check if there is any firewall issue or something idk?  

While mqtt disconnects frequently at my work, the wifi-esp connection doesn't break that often

**Reply by Bernard314** [reply, marked as answer]
ok i ran some more tests. so i have 2 routers at work. i ran my code on the secondary router this time and faced much fewer disconnect-connect cycles (3 in 2 hours) which i feel is very reliable. also the device reconnected much faster. i believe its only my primary router that sucks after all. i'll change them and see how it goes.

sorry i ever questioned your greatness

**Reply by bertmelis** [reply]
question my greatness?

for me it's a matter of testing different environments and making the client more robust. Cases like yours I find interesting.

**Reply by Bernard314** [reply]
thanks! thats great. i think in a few days I will be able to change the routers (or maybe find an unused one to be used solely for my esp32s, i plan on installing around 5 of them). I'll update you on my progress!
